File: competition-website/lib/portable-rng.ts
/**
 * Mulberry32 PRNG — bit-identical to portable_rng.py
 * so Python↔TS physics parity tests can share seeds.
 */
export class PortableRNG {
    private state: number;

    constructor(seed: number = 0) {
        this.state = seed >>> 0;
    }

    seed(seed: number): void {
        this.state = seed >>> 0;
    }

    nextUint32(): number {
        this.state = (this.state + 0x6d2b79f5) >>> 0;
        let t = this.state;
        t = Math.imul(t ^ (t >>> 15), t | 1) >>> 0;
        t = (t ^ (t + Math.imul(t ^ (t >>> 7), t | 61))) >>> 0;
        return (t ^ (t >>> 14)) >>> 0;
    }

    /** Uniform float in [0, 1). */
    random(): number {
        return this.nextUint32() / 4294967296;
    }

    /** Uniform float in [low, high). */
    uniform(low: number = 0, high: number = 1): number {
        return low + (high - low) * this.random();
    }

    /** Integer in [low, high) — matches numpy.RandomState.randint / PortableRNG.randint. */
    randint(low: number, high: number): number {
        low = Math.trunc(low);
        high = Math.trunc(high);
        if (high <= low) return low;
        const span = high - low;
        return low + Math.floor(this.random() * span);
    }

    choice<T>(seq: T[]): T {
        if (seq.length === 0) throw new Error("Cannot choose from an empty sequence");
        return seq[this.randint(0, seq.length)];
    }
}

File: portable_rng.py
"""Portable seeded RNG shared with the TypeScript sim (Mulberry32).

Matches competition-website/lib/portable-rng.ts bit-for-bit so Python↔TS
physics parity tests can use the same seed and action sequence.
"""

from __future__ import annotations


class PortableRNG:
    """Mulberry32 PRNG — identical algorithm in Python and TypeScript."""

    __slots__ = ("_state",)

    def __init__(self, seed: int = 0):
        self._state = int(seed) & 0xFFFFFFFF

    def seed(self, seed: int) -> None:
        self._state = int(seed) & 0xFFFFFFFF

    def next_uint32(self) -> int:
        self._state = (self._state + 0x6D2B79F5) & 0xFFFFFFFF
        t = self._state
        t = ((t ^ (t >> 15)) * (t | 1)) & 0xFFFFFFFF
        t ^= (t + ((t ^ (t >> 7)) * (t | 61))) & 0xFFFFFFFF
        return (t ^ (t >> 14)) & 0xFFFFFFFF

    def random(self) -> float:
        """Uniform float in [0, 1)."""
        return self.next_uint32() / 4294967296.0

    def uniform(self, low: float = 0.0, high: float = 1.0) -> float:
        """Uniform float in [low, high)."""
        return low + (high - low) * self.random()

    def randint(self, low: int, high: int) -> int:
        """Integer in [low, high) — matches numpy.RandomState.randint."""
        low = int(low)
        high = int(high)
        if high <= low:
            return low
        span = high - low
        return low + int(self.random() * span)

    def choice(self, seq):
        if not seq:
            raise IndexError("Cannot choose from an empty sequence")
        return seq[self.randint(0, len(seq))]
